From f22cfbd19cbadbc9b01733cdd20ed2336b7fcb25 Mon Sep 17 00:00:00 2001
From: Herbert Xu <herbert@gondor.apana.org.au>
Date: Sat, 20 Oct 2007 18:49:31 +0800
Subject: [PATCH] [PARSER] Fix here-doc corruption

The change

	[PARSER] Recognise here-doc delimiters terminated by EOF

introduced a regerssion whereby lines starting with eofmark but are not equal
to eofmark would be corrupted.  This patch fixes it.

Test case:

	cat << _ACEOF
	_ASBOX
	_ACEOF

Old result:

	SASBOX

New result:

	_ASBOX
---
 ChangeLog    |    1 +
 src/parser.c |   11 ++++++-----
 2 files changed, 7 insertions(+), 5 deletions(-)

diff --git a/ChangeLog b/ChangeLog
index a821c32..d50d36c 100644
--- a/ChangeLog
+++ b/ChangeLog
@@ -1,6 +1,7 @@
 2007-10-20  Herbert Xu <herbert@gondor.apana.org.au>
 
 	* Added configure --enable-glob and --enable-fnmatch options.
+	* Fix here-doc corruption.
 
 2007-10-17  Herbert Xu <herbert@gondor.apana.org.au>
 
diff --git a/src/parser.c b/src/parser.c
index 4b8a5fe..3832f0b 100644
--- a/src/parser.c
+++ b/src/parser.c
@@ -1047,16 +1047,17 @@ checkend: {
 		if (c == *eofmark) {
 			if (pfgets(line, sizeof line) != NULL) {
 				char *p, *q;
+				int cc;
 
 				p = line;
 				for (q = eofmark + 1;; p++, q++) {
-					c = *p;
-					if (c == '\n')
-						c = 0;
-					if (!*q || c != *q)
+					cc = *p;
+					if (cc == '\n')
+						cc = 0;
+					if (!*q || cc != *q)
 						break;
 				}
-				if (c == *q) {
+				if (cc == *q) {
 					c = PEOF;
 					plinno++;
 					needprompt = doprompt;
-- 
1.5.3.7

